Spi work

lemonpy opened the pull request.

lemonateclypsium committed:
spi: intel-spi: Split hw and sw sequencing

Right now the only driver for Intel's spi has a DANGEROUS tag for
a bug in the past on certain Lenovo platforms. It was cleared out
that the bug was caused for the spi software sequencing mechanism
and if we only use the driver with the hardware sequencing
capabilities will be much safer[1].

This changes will remove all the software sequencing bits from
the driver and left only the hardware sequencing functionality.
If the software sequencing capabilities are needed, the old driver
can be build using the DANGEROUS option from the menu.

[1] https://lkml.org/lkml/2021/11/11/468

lemonateclypsium committed:
spi: intel-spi: Split hw and sw sequencing

Right now the only driver for Intel's spi has a DANGEROUS tag for
a bug in the past on certain Lenovo platforms. It was cleared out
that the bug was caused for the spi software sequencing mechanism
and if we only use the driver with the hardware sequencing
capabilities will be much safer[1].

This changes will remove all the software sequencing bits from
the driver and left only the hardware sequencing functionality.
If the software sequencing capabilities are needed, the old driver
can be build using the DANGEROUS option from the menu.

[1] https://lkml.org/lkml/2021/11/11/468

* Add the config option to choose between hw sequencing only or
  hw and sw sequencing support

mefff commented on drivers/spi/spi-intel.h:
```diff
@@ -10,6 +10,7 @@
 #define SPI_INTEL_H
 
 #include <linux/platform_data/x86/spi-intel.h>
+#include <linux/types.h>
```
I dont see why you need this include here

mefff commented on include/linux/platform_data/x86/spi-intel.h:
```diff
@@ -9,6 +9,8 @@
 #ifndef SPI_INTEL_PDATA_H
 #define SPI_INTEL_PDATA_H
 
+#include <linux/types.h>
```
Ditto for the include

mefff commented on drivers/spi/spi-intel-swseq.c:
```diff
@@ -0,0 +1,189 @@
+{
+	return false;
+}
+#endif
```
Add a comment mentioning what you closed

hughsie commented on drivers/spi/Kconfig:
```diff
@@ -440,17 +440,60 @@ config SPI_INGENIC
+	select SPI_INTEL
+	help
+	  This enables PCI support for the Intel PCH/PCU SPI controller in
+	  master mode. This controller is present in modern Intel hardware
```
isn't the kernel using something else now when referring to master/slave?

marcosdiazr commented on drivers/spi/Kconfig:
```diff
@@ -440,17 +440,60 @@ config SPI_INGENIC
+	  this driver it is possible to upgrade BIOS directly from Linux.
+	  The driver will use HW sequencing capabilities from the chip.
+
+	  Say N here unless you know what you are doing. Overwriting the
```
Why are we still suggesting to use N? isn't this safe enough?

marcosdiazr replied:
I insist to remove this line

marcosdiazr commented on drivers/spi/spi-intel.c:
```diff
@@ -240,9 +97,12 @@ static void intel_spi_dump_regs(struct intel_spi *ispi)
+		ispi->swseq_erase && ispi->swseq_enabled ? 'S' : 'H');
+
+	if (ispi->swseq_enabled)
+		dev_dbg(ispi->dev, "SW sequencer is disabled for all operations\n");
```
enabled

marcosdiazr commented on drivers/spi/spi-intel-swseq.c:
```diff
@@ -0,0 +1,189 @@
+int handle_swseq_wren(const struct intel_spi *ispi)
+{
+	u16 preop;
+	u8 opcode = SPINOR_OP_WREN;
```
const?

marcosdiazr commented on drivers/spi/spi-intel-swseq.c:
```diff
@@ -0,0 +1,189 @@
+				  INTEL_SPI_TIMEOUT * 1000);
+}
+
+static int intel_spi_opcode_index(const struct intel_spi *ispi, const u8 opcode, const int optype)
```
this is about style but i don't think the consts for opcode and optype are needed

marcosdiazr commented on drivers/spi/spi-intel-swseq.c:
```diff
@@ -0,0 +1,189 @@
+	return 0;
+}
+
+int intel_spi_sw_cycle(const struct intel_spi *ispi, const u8 opcode, const size_t len,
```
the same here for opcode, len and optype

marcosdiazr commented on drivers/spi/spi-intel-swseq.c:
```diff
@@ -0,0 +1,189 @@
+
+int handle_swseq_wren(const struct intel_spi *ispi)
+{
+	log_error_swseq_not_supported(ispi);
```
I don't think you should error on this one, I'd return 0 as the original code did when !spi->swseq_reg.
Also I'd rename this function to handle_wren

marcosdiazr replied:
what about this?

marcosdiazr commented on drivers/spi/spi-intel.c:
```diff
@@ -659,7 +406,11 @@ static int intel_spi_erase(struct intel_spi *ispi,
-	if (ispi->swseq_erase)
+	/*
+	 * If swseq_erase is false, it means that we cannot erase using
+	 * HW sequencer.
```
SW?

marcosdiazr commented on drivers/spi/spi-intel.c:
```diff
@@ -1113,12 +856,19 @@ static int intel_spi_init(struct intel_spi *ispi)
 		return -EINVAL;
 	}
 
+	if ((ispi->swseq_erase || !erase_64k) && !ispi->swseq_enabled)
```
why the erase_64k here?

marcosdiazr replied:
Also my general comment, are we sure what ispi->swseq_reg means? are we sure that when that is true hwseq is fully available?

marcosdiazr commented on drivers/spi/spi-intel.c:
```diff
@@ -1113,12 +856,19 @@ static int intel_spi_init(struct intel_spi *ispi)
 	 */
-	if (ispi->swseq_reg) {
+	if (ispi->swseq_reg && ispi->swseq_enabled) {
 		/* Disable #SMI generation from SW sequencer */
```
in this and in the following change you are keeping the SW sequencing code here.
Whereas for example in mem_op_supported_on_spi_locked you moved the function outside to the sw sequencing code. I think we should be consistent, whether to keep the code and protect it with swseq_enabled or use the functions + swseq_enabled

lemonateclypsium committed:
spi: intel-spi: Split hardware and software sequencing

Right now the only driver for Intel's spi has a DANGEROUS tag for
a bug in the past on certain Lenovo platforms. It was cleared out
that the bug was caused for the spi software sequencing mechanism
and if we only use the driver with the hardware sequencing
capabilities will be much safer[1].

This changes will remove all the software sequencing bits from
the driver and left only the hardware sequencing functionality.
If the software sequencing capabilities are needed, the old driver
can be build using the DANGEROUS option from the menu.

[1] https://lkml.org/lkml/2021/11/11/468

* Add the config option to choose between hardware sequencing only or
hardware and software sequencing support

* Add a new config so you have to choose manually if you want
software sequencing support.

marcosdiazr commented on drivers/spi/Kconfig:
```diff
@@ -469,13 +471,22 @@ config SPI_INTEL_PLATFORM
+	depends on SPI_MEM
+	help
+	  This enables software sequencing functionality to the SPI controller
+
```
Add here the comment about using N here unless you know what you are doing

marcosdiazr commented on drivers/spi/spi-intel-swseq.c:
```diff
@@ -0,0 +1,202 @@
+	return true;
+}
+
+int handle_swseq_wren(struct intel_spi *ispi)
```
const struct?

marcosdiazr commented on drivers/spi/spi-intel.c:
```diff
@@ -1118,18 +861,16 @@ static int intel_spi_init(struct intel_spi *ispi)
-	if (ispi->locked && ispi->sregs) {
+	if (ispi->locked && ispi->sregs && ispi->swseq_enabled) {
 		/*
 		 * BIOS programs allowed opcodes and then locks down the
```
shouldn't you move this block too to swseq?
